### ISSUE TITLE
[JENKINS-42034] Reactive Reference parameter with Formated HTML loses script section when binding to other build parameters

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/active-choices-plugin#672]
<p>Using an Active Choices Reactive Reference Parameter and having it return Formatted HTML with a script portion (onchange, etc...) works great unless the groovy script references other dependent build parameters such as strings, Active Choice parameters, etc...  The returned HTML renders properly, just without the script.  Attached is a (hopefully) very simple job configuration to demonstrate.  </p>

<p>The groovy script for generating the sample test is as follows (DummyVar is a string parameter defined just above this parameter):</p>

<pre>
<span class="code-comment">//Uncomment <span class="code-keyword">this</span> line, causing DummyVar to be referenced, and the script will be lost at runtime.
</span><span class="code-comment">//def a = DummyVar
</span>
def FullHTML = ""
FullHTML += <span class="code-quote">"<head>"</span>
FullHTML += <span class="code-quote">"<script type=<span class="code-quote">'text/javascript'</span>> jQuery(document).on(<span class="code-quote">'ready'</span>, function() {	function updateOutputValue()  { jQuery(<span class="code-quote">'#value'</span>).val(<span class="code-quote">'Booya!'</span>);     };"</span>
FullHTML += <span class="code-quote">" updateOutputValue();"</span>
FullHTML += <span class="code-quote">"});"</span>
FullHTML += <span class="code-quote">"</script>"</span>
FullHTML += <span class="code-quote">"</head>"</span>
FullHTML += <span class="code-quote">"<body><form><input id=<span class="code-quote">'value'</span> name=<span class="code-quote">'value'</span> > </form></body>"</span>
<span class="code-keyword">return</span> FullHTML
</pre>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bmatzen">bmatzen</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034" target="_blank">Reactive Reference parameter with Formated HTML loses script section when binding to other build parameters</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ioannis.png" title="ioannis's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ioannis">ioannis</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>component(s)</b>: active-choices-plugin
<li><b>resolution</b>: Not A Defect
<li><b>resolved</b>: 2017-03-11T22:30:34+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 4
<li><b>imported</b>: 2025-12-08
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Using an Active Choices Reactive Reference Parameter and having it return Formatted HTML with a script portion (onchange, etc...) works great unless the groovy script references other dependent build parameters such as strings, Active Choice parameters, etc...  The returned HTML renders properly, just without the script.  Attached is a (hopefully) very simple job configuration to demonstrate.  </p>

<p>The groovy script for generating the sample test is as follows (DummyVar is a string parameter defined just above this parameter):</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java"><span class="code-comment">//Uncomment <span class="code-keyword">this</span> line, causing DummyVar to be referenced, and the script will be lost at runtime.
</span><span class="code-comment">//def a = DummyVar
</span>
def FullHTML = ""
FullHTML += <span class="code-quote">"&lt;head&gt;"</span>
FullHTML += <span class="code-quote">"&lt;script type=<span class="code-quote">'text/javascript'</span>&gt; jQuery(document).on(<span class="code-quote">'ready'</span>, function() {	function updateOutputValue()  { jQuery(<span class="code-quote">'#value'</span>).val(<span class="code-quote">'Booya!'</span>);     };"</span>
FullHTML += <span class="code-quote">" updateOutputValue();"</span>
FullHTML += <span class="code-quote">"});"</span>
FullHTML += <span class="code-quote">"&lt;/script&gt;"</span>
FullHTML += <span class="code-quote">"&lt;/head&gt;"</span>
FullHTML += <span class="code-quote">"&lt;body&gt;&lt;form&gt;&lt;input id=<span class="code-quote">'value'</span> name=<span class="code-quote">'value'</span> &gt; &lt;/form&gt;&lt;/body&gt;"</span>
<span class="code-keyword">return</span> FullHTML
</pre>
</div></div></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Jenkins 2.32.2, Active Choices 1.5.2, Chrome v. 56</code></li></ul>
<details><summary><i>8 attachments</i></summary>

- [config-5-4-2017.xml](https://issues.jenkins.io/secure/attachment/37565/config-5-4-2017.xml)
- [config-doubleload.xml](https://issues.jenkins.io/secure/attachment/37549/config-doubleload.xml)
- [config.xml](https://issues.jenkins.io/secure/attachment/37403/config.xml)
- [config.xml](https://issues.jenkins.io/secure/attachment/35966/config.xml)
- [config[1].xml](https://issues.jenkins.io/secure/attachment/36404/config%5B1%5D.xml)
- [image-2017-06-01-22-46-45-391.png](https://issues.jenkins.io/secure/attachment/38201/image-2017-06-01-22-46-45-391.png)
  > ![image-2017-06-01-22-46-45-391.png](https://issues.jenkins.io/secure/attachment/38201/image-2017-06-01-22-46-45-391.png)
- [image-2017-06-01-22-48-32-552.png](https://issues.jenkins.io/secure/attachment/38202/image-2017-06-01-22-48-32-552.png)
  > ![image-2017-06-01-22-48-32-552.png](https://issues.jenkins.io/secure/attachment/38202/image-2017-06-01-22-48-32-552.png)
- [screenshot-1.png](https://issues.jenkins.io/secure/attachment/36405/screenshot-1.png)
  > ![screenshot-1.png](https://issues.jenkins.io/secure/attachment/36405/screenshot-1.png)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=291322&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-291322">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'm not sure referencing the parameter is the problem. Your script is returning HTML with a &lt;script&gt; tag. The plug-in includes the HTML you provided using innerHTML. That may be the problem.</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ioannis" class="user-hover" rel="ioannis">ioannis</a> has way more experience than I using JavaScript with the reactive parameters. Ioannis, could you take a look at this script, please? Would be great if you could confirm it there's a bug in the plug-in behaviour. Looking at the code, I can't really understand what's going on, except that script tag is never evaluated, as far as I can tell.</p>
</pre>
</details>

</ul>
</details>

<p>I'm not sure referencing the parameter is the problem. Your script is returning HTML with a <script> tag. The plug-in includes the HTML you provided using innerHTML. That may be the problem.</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ioannis" class="user-hover" rel="ioannis">ioannis</a> has way more experience than I using JavaScript with the reactive parameters. Ioannis, could you take a look at this script, please? Would be great if you could confirm it there's a bug in the plug-in behaviour. Looking at the code, I can't really understand what's going on, except that script tag is never evaluated, as far as I can tell.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ioannis.png" title="ioannis's avatar" /> ioannis:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=291370&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-291370">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kinow" class="user-hover" rel="kinow">kinow</a> I think this works as expected. I have attached a new config<span class="error">&#91;1&#93;</span>.xml where I simplified the '<b>ScriptVariables</b>' parameter script so I can easily explain what I think it's going on.</p>

<p>Here is the simplified script:</p>

<p> </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">def a = DummyVar

def FullHTML = """
&lt;script type=<span class="code-quote">'text/javascript'</span>&gt; 
jQuery(document).on(<span class="code-quote">'ready'</span>, function() {	function updateOutputValue() { jQuery(<span class="code-quote">'#value'</span>).val(<span class="code-quote">'Booya!'</span>); };
updateOutputValue();
});
&lt;/script&gt;
&lt;input id=<span class="code-quote">'value'</span> name=<span class="code-quote">"value"</span> type=<span class="code-quote">"text"</span> class=<span class="code-quote">"setting-input "</span> value=<span class="code-quote">"$a"</span>&gt;
"""

<span class="code-keyword">return</span> FullHTML
</pre>
</div></div>
<p>The original script assigned a value through the jQuery selector <b>on page load</b>. </p>

<p>That is the key point I think. The <b>page load script only runs when the build form is first loaded</b>. Not when the parameter is refreshed on a cascade event.</p>

<p>So if you are not referencing another parameter the page loads and the script sets the value to 'Booya'. However, if you are referencing another parameter it triggers an update event as soon as the DummyVar is set. Since the original script did not use the DummyVar then the returned &lt;input&gt; value was null.</p>

<p>In this example, where I'm using the DummyVar value you see that it updates the &lt;input&gt; value to the DummyVar value</p>

<p> <span class="image-wrap" style=""><a id="36405_thumb" href="https://issues.jenkins.io/secure/attachment/36405/36405_screenshot-1.png" title="screenshot-1.png" file-preview-type="image" file-preview-id="36405" file-preview-title="screenshot-1.png"><img src="https://issues.jenkins.io/secure/thumbnail/36405/_thumb_36405.png" style="border: 0px solid black" role="presentation"/></a></span> </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kinow" class="user-hover" rel="kinow">kinow</a> I think this works as expected. I have attached a new config<span class="error">&#91;1&#93;</span>.xml where I simplified the '<b>ScriptVariables</b>' parameter script so I can easily explain what I think it's going on.</p>

<p>Here is the simplified script:</p>

<p> </p>

<pre>
def a = DummyVar

def FullHTML = """
<script type=<span class="code-quote">'text/javascript'</span>> 
jQuery(document).on(<span class="code-quote">'ready'</span>, function() {	function updateOutputValue() { jQuery(<span class="code-quote">'#value'</span>).val(<span class="code-quote">'Booya!'</span>); };
updateOutputValue();
});
</script>
<input id=<span class="code-quote">'value'</span> name=<span class="code-quote">"value"</span> type=<span class="code-quote">"text"</span> class=<span class="code-quote">"setting-input "</span> value=<span class="code-quote">"$a"</span>>
"""

<span class="code-keyword">return</span> FullHTML
</pre>
<p>The original script assigned a value through the jQuery selector <b>on page load</b>. </p>

<p>That is the key point I think. The <b>page load script only runs when the build form is first loaded</b>. Not when the parameter is refreshed on a cascade event.</p>

<p>So if you are not referencing another parameter the page loads and the script sets the value to 'Booya'. However, if you are referencing another parameter it triggers an update event as soon as the DummyVar is set. Since the original script did not use the DummyVar then the returned <input> value was null.</p>

<p>In this example, where I'm using the DummyVar value you see that it updates the <input> value to the DummyVar value</p>

<p> <span class="image-wrap" style=""><a id="36405_thumb" href="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/36405/screenshot-1.png" title="screenshot-1.png" file-preview-type="image" file-preview-id="36405" file-preview-title="screenshot-1.png"><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/36405/screenshot-1.png" style="border: 0px solid black" role="presentation"/></a></span> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=291372&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-291372">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Great explanation - as always - Ioannis. Thanks for taking a look at this issue. I think we can mark it as <del>Won't Fix</del> (you were faster than me <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> ), and wait for <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bmatzen" class="user-hover" rel="bmatzen">bmatzen</a>'s feedback.</p>

<p>ps: Releasing 1.5.3 to update centre in 15 minutes</p>
</pre>
</details>

</ul>
</details>

<p>Great explanation - as always - Ioannis. Thanks for taking a look at this issue. I think we can mark it as <del>Won't Fix</del> (you were faster than me <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> ), and wait for <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bmatzen" class="user-hover" rel="bmatzen">bmatzen</a>'s feedback.</p>

<p>ps: Releasing 1.5.3 to update centre in 15 minutes</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>bmatzen:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=291650&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-291650">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I really appreciate you guys investigating this. I absolutely love this plugin and really get the most out of it, but I'm also pretty novice with Javascript, so I appreciate the help differentiating between Javascript issues versus issues with the actual plugin. BTW I appreciate you pointing out the simplified triple quote syntax for the HTML - much simpler! <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>I've reviewed your feedback, and see that while your solution handles the sample I gave, my sample was pretty stripped down from what I'm really doing, and I need to make an update to help me do what I really want to do.  I consulted a co-worker who is much more savvy with Javascript, and have come up with a modified version of your script above, which I would have expected would work, but is not.</p>

<p>def FullHTML = """
 &lt;script type='text/javascript'&gt; 
 function updateOutputValue(NewVal) { 
     jQuery('#value').val(NewVal); 
 };
 // Handler for the document loading
 jQuery(document).on('ready', function() { 
     alert( "Handler for .ready() begin." ); 
     updateOutputValue('We are live');
     alert( "Handler for .ready() end." );
 });
 // Handler for the 'setting-input' class, of which DummyVar is a part.
 jQuery('.setting-input').change(function() { 
     alert( "Handler for setting-input .change() begin." ); 
     updateOutputValue('DummyVar got updated'); 
     alert( "Handler for setting-input .change() end." );
 });
 &lt;/script&gt;
 &lt;input id='value' name='value' type='text' class='setting-input' value='default value'&gt;
 """
 return FullHTML</p>

<p> </p>

<p>What I'm expecting to happen is for 'value' to first have 'default value', then get updated to 'We are live'. This is working properly.  </p>

<p>Then if/when DummyVar changes (and I leave the text field), I'd expect 'value' to get 'DummyVar got updated'.  <del>Instead what I'm seeing is, although I get the Javascript alerts for changes to '.setting-input' class, instead of 'value' changing to 'DummyVar got updated', or even staying the same (no effect), it's reverting to 'default value'.</del>  Although I get the Javascript alerts for changes to '.setting-input' class, and see 'value' flash to 'DummyVar got updated', it's immediately reverting to 'default value'.</p>

<p>Alternatively if I remove the default value (i.e. have "&lt;input id='value' name='value' type='text'&gt;"), changes to DummyVar cause 'value' to clear out.  So either way it's as though the control gets reloaded. </p>

<p>Hopefully it's clear to you guys what's happening here - quite likely the issue has to do with my script and the way Active Choices interacts with it - but I'm hoping you can clear things up and with luck steer me to some changes to my script to resolve this.</p>

<p>Note that when I use this script outside of Jenkins, simply in its own HTML file, 'value' updates properly per my expectations. Here is that script:</p>
<hr />
<p>&lt;head&gt;
     &lt;input id='DummyVal' name='DummyVal' type='text' class='setting-input' value='asdf'&gt;&lt;br&gt;
     &lt;input id='value' name='value' type='text' class='setting-input' value='default value'&gt;
     &lt;script src='http://ajax.googleapis.com/ajax/libs/jquery/1.7.1/jquery.min.js' type='text/javascript'&gt;&lt;/script&gt;
     &lt;script type='text/javascript'&gt; 
         function updateOutputValue(NewVal) { 
             jQuery('#value').val(NewVal); 
         };
         jQuery(document).on('ready', function() { 
             alert( "Handler for .ready() begin." ); 
             updateOutputValue('We are live');
             alert( "Handler for .ready() end." );
         });
         jQuery('.setting-input').change(function() { 
             alert( "Handler for setting-input .change() begin." ); 
             updateOutputValue('DummyVar got updated'); 
             alert( "Handler for setting-input .change() end." );
         });
     &lt;/script&gt;
 &lt;/head&gt;
 &lt;/body&gt;</p>
<hr />
<p>Thanks again for investigating, and again I'm so appreciative of this plugin. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>Thanks,
 Bill</p>
</pre>
</details>

</ul>
</details>

<p>I really appreciate you guys investigating this. I absolutely love this plugin and really get the most out of it, but I'm also pretty novice with Javascript, so I appreciate the help differentiating between Javascript issues versus issues with the actual plugin. BTW I appreciate you pointing out the simplified triple quote syntax for the HTML - much simpler! <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>I've reviewed your feedback, and see that while your solution handles the sample I gave, my sample was pretty stripped down from what I'm really doing, and I need to make an update to help me do what I really want to do.  I consulted a co-worker who is much more savvy with Javascript, and have come up with a modified version of your script above, which I would have expected would work, but is not.</p>

<p>def FullHTML = """<br/>
 <script type='text/javascript'> <br/>
 function updateOutputValue(NewVal) { <br/>
     jQuery('#value').val(NewVal); <br/>
 };<br/>
 // Handler for the document loading<br/>
 jQuery(document).on('ready', function() { <br/>
     alert( "Handler for .ready() begin." ); <br/>
     updateOutputValue('We are live');<br/>
     alert( "Handler for .ready() end." );<br/>
 });<br/>
 // Handler for the 'setting-input' class, of which DummyVar is a part.<br/>
 jQuery('.setting-input').change(function() { <br/>
     alert( "Handler for setting-input .change() begin." ); <br/>
     updateOutputValue('DummyVar got updated'); <br/>
     alert( "Handler for setting-input .change() end." );<br/>
 });<br/>
 </script><br/>
 <input id='value' name='value' type='text' class='setting-input' value='default value'><br/>
 """<br/>
 return FullHTML</p>

<p> </p>

<p>What I'm expecting to happen is for 'value' to first have 'default value', then get updated to 'We are live'. This is working properly.  </p>

<p>Then if/when DummyVar changes (and I leave the text field), I'd expect 'value' to get 'DummyVar got updated'.  <del>Instead what I'm seeing is, although I get the Javascript alerts for changes to '.setting-input' class, instead of 'value' changing to 'DummyVar got updated', or even staying the same (no effect), it's reverting to 'default value'.</del>  Although I get the Javascript alerts for changes to '.setting-input' class, and see 'value' flash to 'DummyVar got updated', it's immediately reverting to 'default value'.</p>

<p>Alternatively if I remove the default value (i.e. have "<input id='value' name='value' type='text'>"), changes to DummyVar cause 'value' to clear out.  So either way it's as though the control gets reloaded. </p>

<p>Hopefully it's clear to you guys what's happening here - quite likely the issue has to do with my script and the way Active Choices interacts with it - but I'm hoping you can clear things up and with luck steer me to some changes to my script to resolve this.</p>

<p>Note that when I use this script outside of Jenkins, simply in its own HTML file, 'value' updates properly per my expectations. Here is that script:</p>
<hr />
<p><head><br/>
     <input id='DummyVal' name='DummyVal' type='text' class='setting-input' value='asdf'><br><br/>
     <input id='value' name='value' type='text' class='setting-input' value='default value'><br/>
     <script src='http://ajax.googleapis.com/ajax/libs/jquery/1.7.1/jquery.min.js' type='text/javascript'></script><br/>
     <script type='text/javascript'> <br/>
         function updateOutputValue(NewVal) { <br/>
             jQuery('#value').val(NewVal); <br/>
         };<br/>
         jQuery(document).on('ready', function() { <br/>
             alert( "Handler for .ready() begin." ); <br/>
             updateOutputValue('We are live');<br/>
             alert( "Handler for .ready() end." );<br/>
         });<br/>
         jQuery('.setting-input').change(function() { <br/>
             alert( "Handler for setting-input .change() begin." ); <br/>
             updateOutputValue('DummyVar got updated'); <br/>
             alert( "Handler for setting-input .change() end." );<br/>
         });<br/>
     </script><br/>
 </head><br/>
 </body></p>
<hr />
<p>Thanks again for investigating, and again I'm so appreciative of this plugin. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>Thanks,<br/>
 Bill</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ioannis.png" title="ioannis's avatar" /> ioannis:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=291940&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-291940">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Currently away for business. I'll review it when I get back. </p>
</pre>
</details>

</ul>
</details>

<p>Currently away for business. I'll review it when I get back. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>bmatzen:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=292156&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-292156">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>It's very much appreciated, thank you Ioannis.</p>
</pre>
</details>

</ul>
</details>

<p>It's very much appreciated, thank you Ioannis.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>bmatzen:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=296468&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-296468">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi Ioannis, I was wondering if you'd had a chance to review this, and could re-open this issue.  We heavily rely on this plugin, and having this functionality resolved (whether it's a change/fix to the plugin or a change to my Javscript if perhaps there's something I can do on my end) would allow us to move forward with some powerful user interfaces, but we're currently blocked on this.  Thank you for all of your help so far, and this great plugin. </p>

<p> </p>

<p>BTW, I'm happy to create a new ticket that better describes the issue, has a better title, etc, if that helps.</p>

<p> </p>

<p>Bill</p>
</pre>
</details>

</ul>
</details>

<p>Hi Ioannis, I was wondering if you'd had a chance to review this, and could re-open this issue.  We heavily rely on this plugin, and having this functionality resolved (whether it's a change/fix to the plugin or a change to my Javscript if perhaps there's something I can do on my end) would allow us to move forward with some powerful user interfaces, but we're currently blocked on this.  Thank you for all of your help so far, and this great plugin. </p>

<p> </p>

<p>BTW, I'm happy to create a new ticket that better describes the issue, has a better title, etc, if that helps.</p>

<p> </p>

<p>Bill</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ioannis.png" title="ioannis's avatar" /> ioannis:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=296657&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-296657">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bmatzen" class="user-hover" rel="bmatzen">bmatzen</a>; It would be much easier for me to review this if you can provide an example project configuration file. Strip it to the bare minimum to demonstrate the problem and your expectation.</p>

<p>A couple things to keep in mind when you are working with an Active-Choices parameter that returns formatted HTML that contains also javascript.</p>
<ol>
	<li>Separate javascript from HTML. Create an AC-Ref with just the &lt;script&gt; tag. This should not reference any other parameters and it's just used to load the javascript on the page</li>
	<li>I never use head and body elements. These are loaded by the Jenkins build form so I don't know what effect they have</li>
	<li>Make sure that the HTML input elements displayed with an AC-Reference parameter have an attribute name="value" if you want their value returned by the AC parameter</li>
	<li>You can actively change such values via javascript but make sure that you trigger a change event so that the new values get propagated to downstream AC parameters that reference the one that was changed. See: also <a href="https://issues.jenkins-ci.org/browse/JENKINS-33326" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-33326</a></li>
</ol>


<p> </p>
</pre>
</details>

</ul>
</details>

<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bmatzen" class="user-hover" rel="bmatzen">bmatzen</a>; It would be much easier for me to review this if you can provide an example project configuration file. Strip it to the bare minimum to demonstrate the problem and your expectation.</p>

<p>A couple things to keep in mind when you are working with an Active-Choices parameter that returns formatted HTML that contains also javascript.</p>
<ol>
	<li>Separate javascript from HTML. Create an AC-Ref with just the <script> tag. This should not reference any other parameters and it's just used to load the javascript on the page</li>
	<li>I never use head and body elements. These are loaded by the Jenkins build form so I don't know what effect they have</li>
	<li>Make sure that the HTML input elements displayed with an AC-Reference parameter have an attribute name="value" if you want their value returned by the AC parameter</li>
	<li>You can actively change such values via javascript but make sure that you trigger a change event so that the new values get propagated to downstream AC parameters that reference the one that was changed. See: also <a href="https://issues.jenkins-ci.org/browse/JENKINS-33326" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-33326</a></li>
</ol>


<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>bmatzen:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=296831&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-296831">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thank you Ioannis, I've just attached config.xml per your request.  Hopefully this demonstrates the issue.  I reviewed your list of things to keep in mind - I believe I'm following 1-3 properly but please correct me if I'm wrong. Number 4 was interesting, but I don't believe it solves the issue. Regardless of whether I manually update the first (referenced) text field (named 'DummyVar' in my example project) or use the .trigger('change') event, the behavior is the same. After the alert popups for the change handler, the referencing (downstream) parameter first flashes to the expected value ('DummyVar got updated') <em><b>but then immediately changes back to the default from the HTML definition</b></em> ('default value').  </p>

<p>I could really use some help on this one. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>Thanks,
Bill</p>
</pre>
</details>

</ul>
</details>

<p>Thank you Ioannis, I've just attached config.xml per your request.  Hopefully this demonstrates the issue.  I reviewed your list of things to keep in mind - I believe I'm following 1-3 properly but please correct me if I'm wrong. Number 4 was interesting, but I don't believe it solves the issue. Regardless of whether I manually update the first (referenced) text field (named 'DummyVar' in my example project) or use the .trigger('change') event, the behavior is the same. After the alert popups for the change handler, the referencing (downstream) parameter first flashes to the expected value ('DummyVar got updated') <em><b>but then immediately changes back to the default from the HTML definition</b></em> ('default value').  </p>

<p>I could really use some help on this one. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>Thanks,<br/>
Bill</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ioannis.png" title="ioannis's avatar" /> ioannis:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=297615&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-297615">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi Bill;</p>

<p>You are indeed violating rule #1 since the javascript and the &lt;input&gt; element that provides the value to your parameter are not separated but reside in the same AC Reactive Ref (ScriptVariables). As a result the javascript is reloaded when your DummyVar changes. I believe that what you are observing are:</p>
<ol>
	<li>First the action from the loaded javascript that flashes what you expect</li>
	<li>Second the action from the update of the ScriptVariables parameter that
	<ol>
		<li>reloads the java script</li>
		<li>sets the input value to 'default value'</li>
	</ol>
	</li>
</ol>


<p> </p>
</pre>
</details>

</ul>
</details>

<p>Hi Bill;</p>

<p>You are indeed violating rule #1 since the javascript and the <input> element that provides the value to your parameter are not separated but reside in the same AC Reactive Ref (ScriptVariables). As a result the javascript is reloaded when your DummyVar changes. I believe that what you are observing are:</p>
<ol>
	<li>First the action from the loaded javascript that flashes what you expect</li>
	<li>Second the action from the update of the ScriptVariables parameter that
	<ol>
		<li>reloads the java script</li>
		<li>sets the input value to 'default value'</li>
	</ol>
	</li>
</ol>


<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>bmatzen:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=298105&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-298105">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I believe I have narrowed down what is happening here, and have simplified some of the "noise" so hopefully that helps. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>What I'm seeing is that the groovy script for each AC-Reactive Reference is run TWICE. Once where the referenced parameters DO NOT exist (i.e. binding.variables.containsKey() returns false for referenced parameters) and a second time where the referenced parameters DO exist. Although I return Javascript both times (at least the code tries to), I've found that the Javascript I return the first is what gets used.</p>

<p>I've attached (config-doubleload.xml) a sample job ("ScriptIssue") that demonstrates this.  I return Javascript that has a a document "ready" handler that simply does an alert popup. The text from the alert indicates whether it was from the first load of the parameter or the second load (based on whether the referenced parameter - "StringParameter" - existed). Again what I'm seeing is it's the first load's returned Javascript that gets used (so the popup shows "GroovyScript1's default handler for .ready()", as opposed to "GroovyScript1's updated handler for .ready()").  I've added a second build parameter (GroovyScript2) of the same type with the same script except it doesn't directly reference "StringParameter".  Nevertheless it behaves the same, so it doesn't actually have anything to do with the dependency, but simply the first time the script is run - that's the Javascript that gets used.</p>

<p>I included one more scenario - parameter "GroovyScript3" - which DOES NOT return any Javascript the first time, but does the second time. In this case, the Javascript returned doesn't seem to take affect (no alert for it). So again it follows the trend where only the Javascript returned when the build parameter's groovy script is first run that is used.</p>

<p>I have println statements in the groovy scripts of the attached config file, and these of course print to jenkins.out.log, where this is what gets printed:</p>

<p>    Loading parameter for GroovyScript1: StringParameter <b>doesn't exist</b>
    Loading parameter for GroovyScript2: StringParameter <b>doesn't exist</b>
    Loading parameter for GroovyScript3: StringParameter <b>doesn't exist (not returning any Javascript)</b>
    Loading parameter for GroovyScript1: StringParameter <b>exists</b>
    Loading parameter for GroovyScript2: StringParameter <b>doesn't exist</b>
    Loading parameter for GroovyScript3: StringParameter <b>exists</b></p>

<p>I'm pretty certain this is the root cause of what I've been fighting with for all these months - hopefully you guys can confirm this, or straighten me out if I'm doing something wrong. ** Thanks for your help on this!</p>

<p> </p>

<p>Thanks,
Bill</p>
</pre>
</details>

</ul>
</details>

<p>I believe I have narrowed down what is happening here, and have simplified some of the "noise" so hopefully that helps. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>What I'm seeing is that the groovy script for each AC-Reactive Reference is run TWICE. Once where the referenced parameters DO NOT exist (i.e. binding.variables.containsKey() returns false for referenced parameters) and a second time where the referenced parameters DO exist. Although I return Javascript both times (at least the code tries to), I've found that the Javascript I return the first is what gets used.</p>

<p>I've attached (config-doubleload.xml) a sample job ("ScriptIssue") that demonstrates this.  I return Javascript that has a a document "ready" handler that simply does an alert popup. The text from the alert indicates whether it was from the first load of the parameter or the second load (based on whether the referenced parameter - "StringParameter" - existed). Again what I'm seeing is it's the first load's returned Javascript that gets used (so the popup shows "GroovyScript1's default handler for .ready()", as opposed to "GroovyScript1's updated handler for .ready()").  I've added a second build parameter (GroovyScript2) of the same type with the same script except it doesn't directly reference "StringParameter".  Nevertheless it behaves the same, so it doesn't actually have anything to do with the dependency, but simply the first time the script is run - that's the Javascript that gets used.</p>

<p>I included one more scenario - parameter "GroovyScript3" - which DOES NOT return any Javascript the first time, but does the second time. In this case, the Javascript returned doesn't seem to take affect (no alert for it). So again it follows the trend where only the Javascript returned when the build parameter's groovy script is first run that is used.</p>

<p>I have println statements in the groovy scripts of the attached config file, and these of course print to jenkins.out.log, where this is what gets printed:</p>

<p>    Loading parameter for GroovyScript1: StringParameter <b>doesn't exist</b><br/>
    Loading parameter for GroovyScript2: StringParameter <b>doesn't exist</b><br/>
    Loading parameter for GroovyScript3: StringParameter <b>doesn't exist (not returning any Javascript)</b><br/>
    Loading parameter for GroovyScript1: StringParameter <b>exists</b><br/>
    Loading parameter for GroovyScript2: StringParameter <b>doesn't exist</b><br/>
    Loading parameter for GroovyScript3: StringParameter <b>exists</b></p>

<p>I'm pretty certain this is the root cause of what I've been fighting with for all these months - hopefully you guys can confirm this, or straighten me out if I'm doing something wrong. ** Thanks for your help on this!</p>

<p> </p>

<p>Thanks,<br/>
Bill</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>bmatzen:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=298297&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-298297">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I can duplicate this in another environment that has nothing to do with Javascript.  In the attached config-5-4-2017 file, I have a simple job with a string parameter ("RefA") and an AC Reactive Parameter. The AC Reactive Parameter has a groovy script that gets run twice, once where RefA is unknown, and once where it is known, in that order.  I return a list with just one item ("What's RefA") if RefA is unknown, whereas if it is known, I load an XML snippet, and add a value from there into the list that I return.  The logging in jenkins.out.log proves that it's running twice, and in the second time, RefA is known and I am putting a proper value into the list that I return.  Nevertheless the UI shows "What's RefA", which was populated on the first groovy run.</p>

<p>What's more, <b>if I simply comment out the line that adds the XML value to the list (in the second run, where RefA is known), that resulting list is the one the UI shows!</b>  Of course in that case it's an empty dropdown, but adding a fake value also shows.  Note that this appears to be the same issue as <a href="https://issues.jenkins.io/browse/JENKINS-37935" title="Referenced parameters not passed to groovy script for reactive reference" class="issue-link" data-issue-key="JENKINS-37935"><del>JENKINS-37935</del></a>, which was closed a "Cannot reproduce".  </p>

<p>Thanks,
Bill</p>
</pre>
</details>

</ul>
</details>

<p>I can duplicate this in another environment that has nothing to do with Javascript.  In the attached config-5-4-2017 file, I have a simple job with a string parameter ("RefA") and an AC Reactive Parameter. The AC Reactive Parameter has a groovy script that gets run twice, once where RefA is unknown, and once where it is known, in that order.  I return a list with just one item ("What's RefA") if RefA is unknown, whereas if it is known, I load an XML snippet, and add a value from there into the list that I return.  The logging in jenkins.out.log proves that it's running twice, and in the second time, RefA is known and I am putting a proper value into the list that I return.  Nevertheless the UI shows "What's RefA", which was populated on the first groovy run.</p>

<p>What's more, <b>if I simply comment out the line that adds the XML value to the list (in the second run, where RefA is known), that resulting list is the one the UI shows!</b>  Of course in that case it's an empty dropdown, but adding a fake value also shows.  Note that this appears to be the same issue as <a href="https://issue-redirect.jenkins.io/issue/37935" title="Referenced parameters not passed to groovy script for reactive reference" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/37935">JENKINS-37935</a>"><del>JENKINS-37935</del></a>, which was closed a "Cannot reproduce".  </p>

<p>Thanks,<br/>
Bill</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>bmatzen:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=301467&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-301467">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi, I wanted to check in to see if you were able to reproduce this at all. I understand if you haven't been able to get to this, but I was hoping to at least know if you were able to understand the issue better with my latest update. I love this plugin and use it a ton, and am hoping to see this issue resolved (or my code fixed if that's what is needed) to move forward with using it more extensively.</p>

<p>Thanks,
Bill</p>
</pre>
</details>

</ul>
</details>

<p>Hi, I wanted to check in to see if you were able to reproduce this at all. I understand if you haven't been able to get to this, but I was hoping to at least know if you were able to understand the issue better with my latest update. I love this plugin and use it a ton, and am hoping to see this issue resolved (or my code fixed if that's what is needed) to move forward with using it more extensively.</p>

<p>Thanks,<br/>
Bill</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ioannis.png" title="ioannis's avatar" /> ioannis:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=301645&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-301645">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bmatzen" class="user-hover" rel="bmatzen">bmatzen</a> With a small change (highlighted) it all works as expected.</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/38201/38201_image-2017-06-01-22-46-45-391.png" style="border: 0px solid black" /></span></p>

<p>You got to insure that the returned array is an array of Strings (not XML nodes as it was before). Then I get this in the UI</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/38202/38202_image-2017-06-01-22-48-32-552.png" style="border: 0px solid black" /></span></p>

<p>Hope it helps. I'm not sure if the script is run multiple times. I can check with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kinow" class="user-hover" rel="kinow">kinow</a></p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bmatzen" class="user-hover" rel="bmatzen">bmatzen</a> With a small change (highlighted) it all works as expected.</p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/38201/image-2017-06-01-22-46-45-391.png" style="border: 0px solid black" /></span></p>

<p>You got to insure that the returned array is an array of Strings (not XML nodes as it was before). Then I get this in the UI</p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/38202/image-2017-06-01-22-48-32-552.png" style="border: 0px solid black" /></span></p>

<p>Hope it helps. I'm not sure if the script is run multiple times. I can check with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kinow" class="user-hover" rel="kinow">kinow</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>bmatzen:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=305522&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-305522">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi Ioannis, thanks for taking a look, unfortunately this doesn't resolve the issue. As you can see in your output, it's still showing "What's RefA" in the server list, which means it's not working as expected. If you review my last post, I go into more detail.  You mentioned you could check with Bruno about whether the script runs multiple times - would you please do that?  </p>

<p>Thanks,
Bill</p>
</pre>
</details>

</ul>
</details>

<p>Hi Ioannis, thanks for taking a look, unfortunately this doesn't resolve the issue. As you can see in your output, it's still showing "What's RefA" in the server list, which means it's not working as expected. If you review my last post, I go into more detail.  You mentioned you could check with Bruno about whether the script runs multiple times - would you please do that?  </p>

<p>Thanks,<br/>
Bill</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=305922&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-305922">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Not sure what the issue is about exactly. Had a quick look at the issue, but didn't have enough time to read all the comments.</p>

<p>If someone could summarize the issue, maybe I can help? We have 4 small/medium issues fixed already, besides the security issue in Scriptler that is blocking our release.</p>

<p>In the meantime I think we can squeeze a few more issues likes this for the next release <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>Not sure what the issue is about exactly. Had a quick look at the issue, but didn't have enough time to read all the comments.</p>

<p>If someone could summarize the issue, maybe I can help? We have 4 small/medium issues fixed already, besides the security issue in Scriptler that is blocking our release.</p>

<p>In the meantime I think we can squeeze a few more issues likes this for the next release <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>bmatzen:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=305944&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-305944">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I would be so appreciative if you could take a look into this one. I understand the comment thread is pretty long and has evolved a bit from the original message. But if you focus on the one from 5/4/2017 (beginning with "I can duplicate this in another environment that has nothing to do with Javascript..."), that is the most isolated/concise example, and the root issue that is causing me grief. If this could be working around or dare I say "fixed" (if it is truly a defect), we would be able to do some really cool things with this plugin (beyond what we're already doing!). <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>I would be so appreciative if you could take a look into this one. I understand the comment thread is pretty long and has evolved a bit from the original message. But if you focus on the one from 5/4/2017 (beginning with "I can duplicate this in another environment that has nothing to do with Javascript..."), that is the most isolated/concise example, and the root issue that is causing me grief. If this could be working around or dare I say "fixed" (if it is truly a defect), we would be able to do some really cool things with this plugin (beyond what we're already doing!). <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=306222&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-306222">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bmatzen" class="user-hover" rel="bmatzen">bmatzen</a>, thanks a lot for attaching a config.xml file. It saves me minutes, and sometimes hours, trying to reproduce the issue <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p> In less than 5 minutes I successfully reproduced your issue. However, at least in Eclipse I can see a serialization exception, where Jenkins' Stapler (its web framework), or more specifically one of its dependencies responsible for parsing, complains about a certain type of object that it could not understand.</p>

<p> The issue is, from what I could tell, in this part of your active-choices script:</p>

<p><tt>a.b.c.@Name.each</tt>{{{}}
<tt>     p</tt><tt>rintln "Found name '" + it</tt>
<tt>  names &lt;&lt; it</tt><tt>}</tt> </p>

<p>The `it` variable in the loop will be an instance of `groovy.util.slurpersupport.Attribute`. You are adding it to the array names (which doesn't use any type as with generics in Java).</p>

<p>The array of `groovy.util.slurpersupport.Attribute` objects is returned to the client, but the framework fails to serialize them, and so you only see the old value, that was calculated in the first time the parameter was executed.</p>

<p>To solve the issue, you have to make sure to send collections of strings, or elements that the framework will successfully serialize as strings (better stick to always returning strings if possible). Here's a quick way to fix it:</p>

<p><tt>a.b.c.@Name.each</tt>{{{}}
{{       println "Found name '" + it}}
{{       names &lt;&lt; it.text()}}
<tt>}</tt></p>

<p>The `.text()` method is documented <a href="http://docs.groovy-lang.org/latest/html/api/groovy/util/slurpersupport/Attribute.html#text%E2%80%93" class="external-link" target="_blank" rel="nofollow noopener">here</a>.</p>

<p>Tested on Eclipse IDE, with the latest version from git master branch, but I believe it should work with 1.5.3 as well. Thanks a lot for your patience, and let us know if that works!</p>

<p>Cheers</p>

<p>Bruno</p>
</pre>
</details>

</ul>
</details>

<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bmatzen" class="user-hover" rel="bmatzen">bmatzen</a>, thanks a lot for attaching a config.xml file. It saves me minutes, and sometimes hours, trying to reproduce the issue <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p> In less than 5 minutes I successfully reproduced your issue. However, at least in Eclipse I can see a serialization exception, where Jenkins' Stapler (its web framework), or more specifically one of its dependencies responsible for parsing, complains about a certain type of object that it could not understand.</p>

<p> The issue is, from what I could tell, in this part of your active-choices script:</p>

<p><tt>a.b.c.@​Name.each</tt>{{{}}<br/>
<tt>     p</tt><tt>rintln "Found name '" + it</tt><br/>
<tt>  names << it</tt><tt>}</tt> </p>

<p>The `it` variable in the loop will be an instance of `groovy.util.slurpersupport.Attribute`. You are adding it to the array names (which doesn't use any type as with generics in Java).</p>

<p>The array of `groovy.util.slurpersupport.Attribute` objects is returned to the client, but the framework fails to serialize them, and so you only see the old value, that was calculated in the first time the parameter was executed.</p>

<p>To solve the issue, you have to make sure to send collections of strings, or elements that the framework will successfully serialize as strings (better stick to always returning strings if possible). Here's a quick way to fix it:</p>

<p><tt>a.b.c.@​Name.each</tt>{{{}}<br/>
{{       println "Found name '" + it}}<br/>
{{       names << it.text()}}<br/>
<tt>}</tt></p>

<p>The `.text()` method is documented <a href="http://docs.groovy-lang.org/latest/html/api/groovy/util/slurpersupport/Attribute.html#text%E2%80%93" class="external-link" target="_blank" rel="nofollow noopener">here</a>.</p>

<p>Tested on Eclipse IDE, with the latest version from git master branch, but I believe it should work with 1.5.3 as well. Thanks a lot for your patience, and let us know if that works!</p>

<p>Cheers</p>

<p>Bruno</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=306223&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-306223">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>JIRA is not happy with my formatting, sorry. Bill, hope you know which part of the code I'm referring to. Let me know if you want me to upload my updated config.xml. Ignore the extra { }`s that JIRA keeps adding to my preformatted blocks (easier to work on Java code than on this JIRA editor <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> )</p>
</pre>
</details>

</ul>
</details>

<p>JIRA is not happy with my formatting, sorry. Bill, hope you know which part of the code I'm referring to. Let me know if you want me to upload my updated config.xml. Ignore the extra { }`s that JIRA keeps adding to my preformatted blocks (easier to work on Java code than on this JIRA editor <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> )</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>bmatzen:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=306340&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-306340">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi Bruno, thank you for your investigation into this - this does in fact resolve that sample. and I'll watch out for those types of issues in the future. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/>
The prior example (2017-05-03 16:47), however, remains an issue, and appears unrelated to this. This does go back to Javascript so a little more complex, but if you wouldn't mind focusing in on that comment / attachment, that remains a sticking point for me. If you could also touch on the "double load" behavior - that strongly plays into this - that would be informative.  Again, thank you so much for this plugin and for investigation my issues. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>Thanks,
Bill</p>
</pre>
</details>

</ul>
</details>

<p>Hi Bruno, thank you for your investigation into this - this does in fact resolve that sample. and I'll watch out for those types of issues in the future. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/><br/>
The prior example (2017-05-03 16:47), however, remains an issue, and appears unrelated to this. This does go back to Javascript so a little more complex, but if you wouldn't mind focusing in on that comment / attachment, that remains a sticking point for me. If you could also touch on the "double load" behavior - that strongly plays into this - that would be informative.  Again, thank you so much for this plugin and for investigation my issues. <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>Thanks,<br/>
Bill</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=306414&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-306414">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bmatzen" class="user-hover" rel="bmatzen">bmatzen</a>, glad it worked.</p>

<p>Searching for 2017-05-03 on this ticket doesn't return anything for me. In your previous comments, only when you quoted a part of the text I could find the attachment.</p>

<p>I *<b>think</b>* JIRA takes into consideration each user's time zone when displaying comments and attachments. As I live in a weird time zone, probably what I see is different than most others here <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> Could you point to a file name, or quote the beginning of the comment, please, Bill?</p>

<p>I will try to chase what's going on in the JavaScript layer. I am aware of parameters being evaluated twice. But as it is normally harmless, and I never found a simple explanation, I've been postponing working on this. Maybe now we can spend some time investigating it together.</p>

<p>Thanks and sorry for the mess with time zones &amp; comments.</p>

<p>Bruno</p>
</pre>
</details>

</ul>
</details>

<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bmatzen" class="user-hover" rel="bmatzen">bmatzen</a>, glad it worked.</p>

<p>Searching for 2017-05-03 on this ticket doesn't return anything for me. In your previous comments, only when you quoted a part of the text I could find the attachment.</p>

<p>I *<b>think</b>* JIRA takes into consideration each user's time zone when displaying comments and attachments. As I live in a weird time zone, probably what I see is different than most others here <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> Could you point to a file name, or quote the beginning of the comment, please, Bill?</p>

<p>I will try to chase what's going on in the JavaScript layer. I am aware of parameters being evaluated twice. But as it is normally harmless, and I never found a simple explanation, I've been postponing working on this. Maybe now we can spend some time investigating it together.</p>

<p>Thanks and sorry for the mess with time zones & comments.</p>

<p>Bruno</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>bmatzen:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=306416&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-306416">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Sorry about that Bruno - ya, the one that starts out "I believe I have narrowed down what is happening here" and references "config-doubleload.xml"... <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>Thanks,
Bill</p>
</pre>
</details>

</ul>
</details>

<p>Sorry about that Bruno - ya, the one that starts out "I believe I have narrowed down what is happening here" and references "config-doubleload.xml"... <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>Thanks,<br/>
Bill</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=306417&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-306417">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Not a problem, and thanks for the prompt reply.</p>

<p>Here's what is says for me on that comment: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bmatzen" class="user-hover" rel="bmatzen">bmatzen</a> added a comment - 2017-05-04 11:47, so I believe JIRA is indeed storing some UTC value and then rendering the currentUser() tz time. Would be good to have a comment ID somewhere I think... sorry for digressing.</p>

<p>Will update the ticket once I have time to investigate it (bit busy with the biouno website today)</p>

<p>Bruno</p>
</pre>
</details>

</ul>
</details>

<p>Not a problem, and thanks for the prompt reply.</p>

<p>Here's what is says for me on that comment: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bmatzen" class="user-hover" rel="bmatzen">bmatzen</a> added a comment - 2017-05-04 11:47, so I believe JIRA is indeed storing some UTC value and then rendering the currentUser() tz time. Would be good to have a comment ID somewhere I think... sorry for digressing.</p>

<p>Will update the ticket once I have time to investigate it (bit busy with the biouno website today)</p>

<p>Bruno</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kinow.png" title="kinow's avatar" /> kinow:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=306833&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-306833">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hmmm, did some initial analysis, and there are a few different ways to fix it. I <em>think</em> the scripts are evaluated many times because we have, for the <a href="https://github.com/jenkinsci/active-choices-plugin/blob/978eac9e2ef913b1e3ae93b0d61ed7245e60d597/src/main/resources/org/biouno/unochoice/CascadeChoiceParameter/index.jelly#L72" class="external-link" target="_blank" rel="nofollow noopener">cascade parameters</a> and for <a href="https://github.com/jenkinsci/active-choices-plugin/blob/978eac9e2ef913b1e3ae93b0d61ed7245e60d597/src/main/resources/org/biouno/unochoice/DynamicReferenceParameter/index.jelly#L132" class="external-link" target="_blank" rel="nofollow noopener">dynamic parameters</a>, they are updated once they are displayed on the UI.</p>

<p>There is an old issue, somewhere in JIRA (bad memory, sorry) where users had to click somewhere to active check boxes or radio buttons. It is tricky to control the DOM creation and the dynamic parameters. So it was the easiest solution at that time (I think it was after a major refactoring in the code, when it was uno-choice, in our own update centre).</p>

<p>Now it looks like we need to properly fix it. I will log an issue for that, linking to this one. Feel free to watch and comment on that issue.</p>

<p>What I have in my mind as a possible and good solution is:</p>
<ul>
	<li>Create a graph/tree of all parameters, hierarchically</li>
	<li>Walk the graph/tree, evaluating each parameter</li>
	<li>Re-use this graph model to evaluate active-choices parameters without an UI</li>
</ul>


<p>This last option has been on my mind for a while. There is another old issue, to support timers/pipelines/cron jobs/etc. But given the current nature of the plug-in, it is impossible to evaluate plug-ins without an UI.</p>

<p>With this graph/tree model, we would solve the issue of evaluating parameters once in the UI (DOM/JS/HTML), and also adding the ability to have new backend parameters (Java only).</p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>Hmmm, did some initial analysis, and there are a few different ways to fix it. I <em>think</em> the scripts are evaluated many times because we have, for the <a href="https://github.com/jenkinsci/active-choices-plugin/blob/978eac9e2ef913b1e3ae93b0d61ed7245e60d597/src/main/resources/org/biouno/unochoice/CascadeChoiceParameter/index.jelly#L72" class="external-link" target="_blank" rel="nofollow noopener">cascade parameters</a> and for <a href="https://github.com/jenkinsci/active-choices-plugin/blob/978eac9e2ef913b1e3ae93b0d61ed7245e60d597/src/main/resources/org/biouno/unochoice/DynamicReferenceParameter/index.jelly#L132" class="external-link" target="_blank" rel="nofollow noopener">dynamic parameters</a>, they are updated once they are displayed on the UI.</p>

<p>There is an old issue, somewhere in JIRA (bad memory, sorry) where users had to click somewhere to active check boxes or radio buttons. It is tricky to control the DOM creation and the dynamic parameters. So it was the easiest solution at that time (I think it was after a major refactoring in the code, when it was uno-choice, in our own update centre).</p>

<p>Now it looks like we need to properly fix it. I will log an issue for that, linking to this one. Feel free to watch and comment on that issue.</p>

<p>What I have in my mind as a possible and good solution is:</p>
<ul>
	<li>Create a graph/tree of all parameters, hierarchically</li>
	<li>Walk the graph/tree, evaluating each parameter</li>
	<li>Re-use this graph model to evaluate active-choices parameters without an UI</li>
</ul>


<p>This last option has been on my mind for a while. There is another old issue, to support timers/pipelines/cron jobs/etc. But given the current nature of the plug-in, it is impossible to evaluate plug-ins without an UI.</p>

<p>With this graph/tree model, we would solve the issue of evaluating parameters once in the UI (DOM/JS/HTML), and also adding the ability to have new backend parameters (Java only).</p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>bmatzen:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-42034?focusedId=308030&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-308030">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This would be great Bruno - I see the new issue (<a href="https://issues.jenkins.io/browse/JENKINS-45507" title="Evaluate parameters once" class="issue-link" data-issue-key="JENKINS-45507">JENKINS-45507</a>) and will monitor that for updates. I'm also more than happy to test anything out before you officially release any update. Thank you so much for investigating this and getting to the root issue.</p>

<p>Thanks,
Bill</p>
</pre>
</details>

</ul>
</details>

<p>This would be great Bruno - I see the new issue (<a href="https://issue-redirect.jenkins.io/issue/45507" title="Evaluate parameters once" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/45507">JENKINS-45507</a>">JENKINS-45507</a>) and will monitor that for updates. I'm also more than happy to test anything out before you officially release any update. Thank you so much for investigating this and getting to the root issue.</p>

<p>Thanks,<br/>
Bill</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a class="jira-relationship" href="https://issue-redirect.jenkins.io/issue/45507">JENKINS-45507</a>]</i>
